tools/pubsys/src/aws/validate_ami/results.rs

```diff
@@ -0,0 +1,714 @@
+//! The results module owns the reporting of EC2 image validation results.
+
+use super::ami::{ImageDef, Result};
+use aws_sdk_ec2::Region;
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, HashSet};
+use std::fmt::{self, Display};
+use std::str::FromStr;
+use tabled::{Table, Tabled};
+
+/// Represent the possible status of an EC2 image validation
+#[derive(Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub(crate) enum AmiValidationResultStatus {
+    /// The expected value was equal to the actual value
+    Correct,
+
+    /// The expected value was different from the actual value
+    Incorrect,
+
+    /// The image was expected but not included in the actual images
+    Missing,
+}
+
+impl Display for AmiValidationResultStatus {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            AmiValidationResultStatus::Correct => write!(f, "Correct"),
+            AmiValidationResultStatus::Incorrect => write!(f, "Incorrect"),
+            AmiValidationResultStatus::Missing => write!(f, "Missing"),
+        }
+    }
+}
+
+impl FromStr for AmiValidationResultStatus {
+    type Err = super::Error;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s {
+            "Correct" => Ok(Self::Correct),
+            "Incorrect" => Ok(Self::Incorrect),
+            "Missing" => Ok(Self::Missing),
+            filter => Err(Self::Err::InvalidStatusFilter {
+                filter: filter.to_string(),
+            }),
+        }
+    }
+}
+
+/// Represents a single EC2 image validation result
+#[derive(Debug, Eq, Hash, PartialEq, Serialize)]
+pub(crate) struct AmiValidationResult {
+    /// The id of the image
+    pub(crate) image_id: String,
+
+    /// The expected value of the image
+    pub(crate) expected_value: ImageDef,
+
+    /// The actual retrieved value of the image
+    pub(crate) actual_value: Option<ImageDef>,
+
+    /// The region the image resides in
+    #[serde(serialize_with = "serialize_region")]
+    pub(crate) region: Region,
+
+    /// The validation status of the image
+    pub(crate) status: AmiValidationResultStatus,
+}
+
+fn serialize_region<S>(region: &Region, serializer: S) -> std::result::Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    serializer.serialize_str(region.to_string().as_str())
+}
+
+impl AmiValidationResult {
+    pub(crate) fn new(
+        image_id: String,
+        expected_value: ImageDef,
+        actual_value: Option<ImageDef>,
+        region: Region,
+    ) -> Self {
+        // Determine the validation status based on equality, presence, and absence of expected and
+        // actual image values
+        let status = match (&expected_value, &actual_value) {
+            (expected_value, Some(actual_value)) if actual_value.eq(expected_value) => {
```

Comment:
slightly more idiomatic to use the operator

```suggestion
            (expected_value, Some(actual_value)) if actual_value. == expected_value => {
```